internal/runner/jest.go

```diff
@@ -0,0 +1,196 @@
+package runner
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"slices"
+	"strings"
+
+	"github.com/buildkite/test-splitter/internal/debug"
+	"github.com/buildkite/test-splitter/internal/plan"
+	"github.com/kballard/go-shellquote"
+)
+
+var _ = TestRunner(Jest{})
+
+type Jest struct {
+	RunnerConfig
+}
+
+func NewJest(j RunnerConfig) Jest {
+	if j.TestCommand == "" {
+		j.TestCommand = "yarn test {{testExamples}} --json --testLocationInResults --outputFile {{outputFile}}"
+	}
+
+	if j.TestFilePattern == "" {
+		j.TestFilePattern = "**/{__tests__/**/*,*.spec,*.test}.{ts,js,tsx,jsx}"
+	}
+
+	if j.TestFileExcludePattern == "" {
+		j.TestFileExcludePattern = "node_modules"
+	}
+
+	if j.RetryTestCommand == "" {
+		j.RetryTestCommand = "yarn test --testNamePattern '{{testNamePattern}}' --json --testLocationInResults --outputFile {{outputFile}}"
+	}
+
+	return Jest{j}
+}
+
+func (j Jest) Name() string {
+	return "Jest"
+}
+
+// GetFiles returns an array of file names using the discovery pattern.
+func (j Jest) GetFiles() ([]string, error) {
+	debug.Println("Discovering test files with include pattern:", j.TestFilePattern, "exclude pattern:", j.TestFileExcludePattern)
+	files, err := discoverTestFiles(j.TestFilePattern, j.TestFileExcludePattern)
+	debug.Println("Discovered", len(files), "files")
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(files) == 0 {
+		return nil, fmt.Errorf("no files found with pattern %q and exclude pattern %q", j.TestFilePattern, j.TestFileExcludePattern)
+	}
+
+	return files, nil
+}
+
+func (j Jest) Run(testCases []string, retry bool) (RunResult, error) {
+	var cmd *exec.Cmd
+	var err error
+
+	defer os.Remove(j.ResultPath)
+
+	if !retry {
+		commandName, commandArgs, err := j.commandNameAndArgs(j.TestCommand, testCases)
+		if err != nil {
+			return RunResult{Status: RunStatusError}, fmt.Errorf("failed to build command: %w", err)
+		}
+
+		cmd = exec.Command(commandName, commandArgs...)
+	} else {
+		commandName, commandArgs, err := j.retryCommandNameAndArgs(j.RetryTestCommand, testCases)
+		if err != nil {
+			return RunResult{Status: RunStatusError}, fmt.Errorf("failed to build command: %w", err)
+		}
+
+		cmd = exec.Command(commandName, commandArgs...)
+	}
+
+	err = runAndForwardSignal(cmd)
+
+	if err == nil {
```

Comment:
returning early when error is nil is unusual in Go, maybe add comment. https://github.com/buildkite/test-splitter/pull/156#discussion_r1721323894
```suggestion
	if err == nil { // note: returning success early
```